### PR TITLE
ready for review

### PR DESCRIPTION
**Pull request by fxqhyw**

### CODE REVIEW

File: app/controllers/checkouts_controller.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+class CheckoutsController < ApplicationController
+  include Wicked::Wizard
+  include CheckoutUpdater
```
Comment (Svatok) [inline app/controllers/checkouts_controller.rb]: Move this logic to services. One service will be like router that will call different services for each step.

File: app/controllers/concerns/order_transfer.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+  included do
+    private
+
+    def transfer_order_to_user
+      order = Order.find_by_id(cookies.signed[:order_id])
+      return unless order
+      @user.orders.in_progress.destroy_all
+      @user.orders << order
+      @user.save
```
Comment (Svatok) [inline app/controllers/concerns/order_transfer.rb]: `order.update(user: @user)` instead this two lines

File: app/controllers/order_items_controller.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
+class OrderItemsController < ApplicationController
+  before_action :order_item, only: %i[update destroy]
+
+  def create
+    create_order unless current_order
```
Comment (Svatok) [inline app/controllers/order_items_controller.rb]: Move all logic to service.

File: app/controllers/registrations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
+class RegistrationsController < Devise::RegistrationsController
+  include OrderTransfer
```
Comment (Svatok) [inline app/controllers/registrations_controller.rb]: Move logic to service

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
+class SessionsController < Devise::SessionsController
+  include OrderTransfer
```
Comment (Svatok) [inline app/controllers/sessions_controller.rb]: Move logic to service

File: app/decorators/order_decorator.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+class OrderDecorator < Draper::Decorator
+  delegate_all
+
+  def status_string
+    case object.status
+    when 'in_queue' then I18n.t('orders.waiting')
```
Comment (Svatok) [inline app/decorators/order_decorator.rb]: Add status key to locale file. For examle:
```yml
orders:
  in_queue: text...
```
Then you can update method like this:
```ruby
def status_string
  I18n.t("orders.#{object.status}")
end
```
Using `case` is bad practice.

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
+module ApplicationHelper
+  def flash_class(level)
+    case level
```
Comment (Svatok) [inline app/helpers/application_helper.rb]: Please use  hash:
```ruby
{
  'notice' => 'alert alert-info',
  ...
}[level]
```

File: app/helpers/books_helper.rb [hunk 8]
```diff
@@ -0,0 +1,33 @@
+module BooksHelper
+  def filter_books_title
+    case request.GET[:filter]
```
Comment (Svatok) [inline app/helpers/books_helper.rb]:
```ruby
def filter_books_title
  I18n.t("button.#{request.GET[:filter]}")
end
```

File: app/helpers/books_helper.rb [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  def secondary_image(image)
+    url_for(image.variant(resize: '150x190!'))
+  end
+
+  def slider_image(book)
+    url_for(book.images.first.variant(resize: '250x316!'))
+  end
+
+  def more_books?(books)
+    books.any? && books.size > 11
```
Comment (Svatok) [inline app/helpers/books_helper.rb]: Move `11` to constant

File: app/helpers/checkouts_helper.rb [hunk 10]
```diff
@@ -0,0 +1,60 @@
+module CheckoutsHelper
+  def active_step(current_step)
+    'active' if current_step == step
+  end
+
+  def address_error?(field:, tag:)
+    return @billing.errors.include?(field) if @billing && tag == 'billing'
+    @shipping.errors.include?(field) if @shipping && tag == 'shipping'
```
Comment (Svatok) [inline app/helpers/checkouts_helper.rb]: Move `@shipping && tag == 'shipping'` to private method. You use this condition in several methods.

File: app/helpers/checkouts_helper.rb [hunk 11]
```diff
@@ -0,0 +1,60 @@
+module CheckoutsHelper
+  def active_step(current_step)
+    'active' if current_step == step
+  end
+
+  def address_error?(field:, tag:)
+    return @billing.errors.include?(field) if @billing && tag == 'billing'
```
Comment (Svatok) [inline app/helpers/checkouts_helper.rb]: Move `@billing && tag == 'billing'` to private method. You use this condition in several methods.

File: app/helpers/checkouts_helper.rb [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+  def card_saved_value(field)
+    current_order.credit_card.try(field) || @credit_card.try(field)
+  end
+
+  def client_name(address)
+    address.first_name + ' ' + address.last_name
+  end
+
+  def secret_card_number(number)
+    '*** *** *** ' + number[-4, 4]
```
Comment (Svatok) [inline app/helpers/checkouts_helper.rb]: Save all numbers of credit card to database is bad practice. Please save only last 4 digets.

File: app/helpers/orders_helper.rb [hunk 13]
```diff
@@ -0,0 +1,10 @@
+module OrdersHelper
+  def filter_orders_title
+    case request.GET[:filter]
```
Comment (Svatok) [inline app/helpers/orders_helper.rb]:
```ruby
def filter_orders_title
  I18n.t("orders.#{request.GET[:filter]}")
end
```

File: app/helpers/reviews_helper.rb [hunk 14]
```diff
@@ -0,0 +1 @@
+module ReviewsHelper; end
```
Comment (Svatok) [inline app/helpers/reviews_helper.rb]: Do you use this helper?

File: app/mailers/checkout_mailer.rb [hunk 15]
```diff
@@ -0,0 +1,9 @@
+class CheckoutMailer < ApplicationMailer
+  default from: 'noreply@amazingbookstore.com'
+
+  def complete_email
+    @user = params[:user]
+    @order = params[:order]
+    mail(to: @user.email, subject: 'Thank You for your Order!')
```
Comment (Svatok) [inline app/mailers/checkout_mailer.rb]: subject to i18

File: app/mailers/registration_mailer.rb [hunk 16]
```diff
@@ -0,0 +1,9 @@
+class RegistrationMailer < ApplicationMailer
+  default from: 'noreply@amazingbookstore.com'
+
+  def registration_email
+    @email = params[:email]
+    @password = params[:password]
+    mail(to: @email, subject: 'Welcome to our Amazing Bookstore!')
```
Comment (Svatok) [inline app/mailers/registration_mailer.rb]: subject to i18

File: app/models/concerns/order_summary.rb [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+    delivery.try(:price) || 0.00
+  end
+
+  def order_total
+    subtotal - discount + delivery_price
+  end
+
+  def items_count
+    count = order_items.collect(&:quantity).compact.sum
+    return false if count.zero?
```
Comment (Svatok) [inline app/models/concerns/order_summary.rb]: I think it is not readable when method `items_count` return false. According to name this method must return integer.

File: app/models/review.rb [hunk 18]
```diff
@@ -0,0 +1,31 @@
+class Review < ApplicationRecord
+  include ReviewerInfo
```
Comment (Svatok) [inline app/models/review.rb]: Move this to decorator.

File: app/services/checkout_stepper.rb [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+      @step = step unless completed?(step)
+    end
+    @step
+  end
+
+  private
+
+  def completed?(step)
+    case step
+    when :login
```
Comment (Svatok) [inline app/services/checkout_stepper.rb]: Please use hash

File: db/migrate/20180716131655_create_credit_cards.rb [hunk 20]
```diff
@@ -0,0 +1,13 @@
+class CreateCreditCards < ActiveRecord::Migration[5.2]
+  def change
+    create_table :credit_cards do |t|
+      t.string :first_name
+      t.string :last_name
+      t.string :number
+      t.string :expiration_date
+      t.string :cvv
```
Comment (Svatok) [inline db/migrate/20180716131655_create_credit_cards.rb:8]: Do not save CVV!!!!!!!!

File: sddfit [hunk 21]
```diff
@@ -0,0 +1,840 @@
+=> [[32m#<Book:0x000000000380fe50[0m
```
Comment (Svatok) [inline sddfit]: Please remove this file or add to gitignore.

**Review by Svatok — CHANGES_REQUESTED** [review]
1. Bug:
When open https://loker-bookstore.herokuapp.com/catalog?category=1 i see 11 books. When i select sorting popular first (https://loker-bookstore.herokuapp.com/catalog?category=1&filter=popular) i see only 2 books. Please fix this and add tests for this case.
2. Bug:
On page https://loker-bookstore.herokuapp.com/checkouts/address when i click `same as billing` i can't unselect `same as billing` if I change my mind. Please fix this and add tests.
3. Bug:
When i have filled billing address and select 'same as billing' for shipping address, then i choose shipping in next step and go back to address page i see empty shipping address. Please fix this and add tests.
4. Bug:
I don't see `cancelled` status on https://loker-bookstore.herokuapp.com/orders page
5. Bug:
I can't reset password with `forgot password`. When i fill input with my email and select `Send me reset password instructions` i see 500 error.
6. Bug:
Wrong button text on https://loker-bookstore.herokuapp.com/users/sign_in page
7. Emails are not sent. I used two mailboxes google and i.ua
8. Facebook login and signup don't work
9. You often use concerns in controllers. Please use services instead of concerns.For example you can use gem that provides architecture for services (for example https://github.com/andypike/rectify) or use your own services.

File: app/services/add_coupon.rb [hunk 22]
```diff
@@ -0,0 +1,22 @@
+class AddCoupon < Rectify::Command
+  def initialize(code:, order:)
+    @code = code
```
Comment (Svatok) [inline app/services/add_coupon.rb:5]: Add `attr_reader` for `code` and `order`

File: app/services/order_transfer.rb [hunk 23]
```diff
@@ -0,0 +1,30 @@
+class OrderTransfer < Rectify::Command
+  def initialize(user:, order_id:)
+    @user = user
+    @order_id = order_id
```
Comment (Svatok) [inline app/services/order_transfer.rb]: Remove this variable and add to initializer `@order = Order.find_by_id(order_id)`.
Also please use `attr_reader` for `@user` and `@order`

**Review by Svatok — CHANGES_REQUESTED** [review]

File: app/services/settings_addresser.rb [hunk 24]
```diff
@@ -1,5 +1,22 @@
-class SettingsAddresser
-  def self.call(params)
-    Address.find_by(user_id: params[:user_id], type: params[:type]) || Address.new(user_id: params[:user_id], type: params[:type])
+class SettingsAddresser < Rectify::Command
+  def initialize(params)
+    @params = params
+  end
```
Comment (Svatok) [inline app/services/settings_addresser.rb:6]: Add variable `@address = find_or_initialize_by(user_id: @params[:user_id], type: @params[:type])` instead `find_address` and `new_address`
Also please use attr_reader for @params and @address

**Review by Svatok — CHANGES_REQUESTED** [review]

File: app/services/create_order_item.rb [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+
+  def find_order_item
+    @order_item = @order.order_items.find_by(book_id: @params[:book_id])
+  end
+
+  def update_order_item_quantity
+    return create_new_order_item unless @order_item
+
+    @order_item.quantity += @params[:quantity].to_i
+    @order_item.save
```
Comment (Svatok) [inline app/services/create_order_item.rb]: @order_item.update(quantity: @order_item.quantity + @params[:quantity].to_i)

File: app/services/create_order_item.rb [hunk 26]
```diff
@@ -0,0 +1,38 @@
+class CreateOrderItem < Rectify::Command
+  def initialize(order:, user:, params:)
+    @order = order
```
Comment (Svatok) [inline app/services/create_order_item.rb]: Please use attr_reader

**Review by Svatok — CHANGES_REQUESTED** [review]

File: app/services/create_order_item.rb [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+
+  def create_order
+    return current_user.orders.create if user_signed_in?
+
+    order = Order.create
+    cookies.signed[:order_id] = { value: order.id, expires: 1.month.from_now }
+  end
+
+  def find_order_item
+    @order_item = current_order.order_items.find_by(book_id: @params[:book_id])
```
Comment (Svatok) [inline app/services/create_order_item.rb]: Move this to initializer

File: app/services/create_order_item.rb [hunk 28]
```diff
@@ -0,0 +1,33 @@
+class CreateOrderItem < Rectify::Command
+  def initialize(params)
+    @params = params
```
Comment (Svatok) [inline app/services/create_order_item.rb:5]: Please use attr_reader

**Review by Svatok — CHANGES_REQUESTED** [review]

**Review by Svatok — CHANGES_REQUESTED** [review]
Please use `attr_reader` where it is possible.

File: app/services/create_order_item.rb (hunk 28 above)
Comment (fxqhyw) [reply app/services/create_order_item.rb:5]: Done

**Review by Svatok — APPROVED** [review]

**Pull request closed without merging**